controller/lib/hal/stm32.h

```diff
@@ -0,0 +1,361 @@
+      REG rsvd3[2];
+      REG iSeq;                  // 0x4C - Injected sequence regiseter
+      REG rsvd4[4];
+      REG offset[4];             // 0x60 - Offset registers
+      REG rsvd5[4];
+      REG iData[4];              // 0x80 - Injected channel data
+      REG rsvd6[4];
+      REG wdCfg[2];              // 0xA0 - Watchdog config
+      REG rsvd7[2];
+      REG diffSel;               // 0xB0 - Differential mode selection
+      REG cal;                   // 0xB4 - Calibration factors
+      REG rsvd8[18];
+   } adc[2];
+
+   // A/D common registers
+   REG comStat;                  // 0x300 - Common status
+   REG rsvd9;
+   REG comCtrl;                  // 0x304 - Common control
+   REG comData;                  // 0x308 - Common data
+};
+
+#define TIMER1_BASE                 0x40012C00
+#define TIMER2_BASE                 0x40000000
+#define TIMER3_BASE                 0x40000400
+#define TIMER6_BASE                 0x40001000
+#define TIMER7_BASE                 0x40001400
+#define TIMER15_BASE                0x40014000
+#define TIMER16_BASE                0x40014400
+struct TimerRegs
+{
+   REG ctrl[2];
+   REG slaveCtrl;
+   REG intEna;
+   REG status;
+   REG event;
+   REG ccMode[2];
+   REG ccEnable;
+   REG counter;
+   REG prescale;
+   REG reload;
+   REG repeat;
+   REG compare[4];
+   REG deadTime;
+   REG dmaCtrl;
+   REG dmaAddr;
+   REG opt1;
+   REG ccMode3;
+   REG compare5;
+   REG compare6;
+   REG opt2;
+   REG opt3;
+};
+
+#define FLASH_BASE                 0x40022000
+struct FlashReg
+{
+   REG access;
+   REG pdKey;
+   REG key;
+   REG optKey;
+   REG status;
+   REG ctrl;
+   REG ecc;
+   REG rsvd1;
+   REG option;
+   REG pcropStart;
+   REG pcropEnd;
+   REG wrpA;
+   REG wrpB;
+};
+
+#define DMA1_BASE                  0x40020000
+#define DMA2_BASE                  0x40020400
+struct DMA_Reg
+{
+   REG intStat;
+   REG intClr;
+   struct
+   {
+      REG config;
+      REG count;
+      REG pAddr;
+      REG mAddr;
+      REG rsvd;
+   } channel[7];
+   REG rsvd[5];
+   REG chanSel;
+};
+
+#define SPI1_BASE                  0x40013000
+#define SPI2_BASE                  0x40003800
+#define SPI3_BASE                  0x40003C00
+struct SPI_Regs
+{
+   REG ctrl[2];
+   REG status;
+   REG data;
+   REG crcPoly;
+   REG rxCRC;
+   REG txCRC;
+};
+
+#define I2C1_BASE                  0x40005400
+#define I2C2_BASE                  0x40005800
+#define I2C3_BASE                  0x40005c00
+#define I2C4_BASE                  0x40008400
+struct I2C_Regs
+{
+   REG ctrl[2];
+   REG addr[2];
+   REG timing;
+   REG timeout;
+   REG status;
+   REG intClr;
+   REG pec;
+   REG rxData;
+   REG txData;
+};
+
+// General Purpose I/O
+#define GPIO_A_BASE                 0x48000000
+#define GPIO_B_BASE                 0x48000400
+#define GPIO_C_BASE                 0x48000800
+#define GPIO_D_BASE                 0x48000C00
+#define GPIO_E_BASE                 0x48001000
+#define GPIO_H_BASE                 0x48001C00
```

Comment:
Yeah, with all of these tables and constants, I keep wondering "how do we know there isn't a typo". I'm almost certain that if I was just hand-copying all of these things from the manual, I would make several typos. Did you use some other method, e.g. copy these from some battle-tested source? If so, please indicate where.

Reply:
The code here was largely copied from my flow sensor project which happens to use a processor in the same family, so has the same peripheral address mapping.  Ultimately though it was initially copied by hand from the reference manual.
Ultimately, you have to test the code to make sure the hardware is doing what you think it should.  I've done some of that testing already and will do more once I get the board that Edwin will likely ship to me today.